proposals/0001-lhls.md

```diff
@@ -0,0 +1,313 @@
+
+### EXT-X-DISCONTINUITY
+[discontinuities]: #ext-x-discontinuity
+
+A prefetch segment must not be advertised with an `EXT-X-DISCONTINUITY` tag. To insert a discontinuity just for prefetch segments, the server must insert the `EXT-X-PREFETCH-DISCONTINUITY` tag before the newest `EXT-X-PREFETCH` tag of the new discontinuous range. 
+
+### EXT-X-MAP
+
+Prefetch segments must not be advertised with an `EXT-X-MAP` tag.
+
+### EXT-X-KEY
+
+Prefetch segments may be advertised with an `EXT-X-KEY` tag. The key itself must be complete; the server must not expect the client to progressively stream keys.
+
+
+## Client/Server Responsibilities
+
+This section amends any language in RFC8216 which would preclude streaming a prefetch segment. The term "Media Segment" also applies to a prefetch segment. As always, unless otherwise countermanded all other requirements apply. 
+
+## General Server Responsibilities
+
+Any Media Segment that is specified in a Playlist loaded by a client MUST be available for immediate download unless it has been marked with an EXT-X-PREFETCH or EXT-X-GAP tag; otherwise, playback errors can occur. Once download starts, only prefetch segments may have their transfer rate constrained by the segment production process.
+
+In addition to the outlined cases, the server must not change the Media Playlist file except to:
+
+* Transform a prefetch segment to a complete segment. ([Prefetch Transformation](#prefetch-transformation))
+
+To each prefetch segment response, the server must append the `Transfer-Encoding: chunked` header. The server must maintain the persistent HTTP connection long enough for a client to receive the entire segment - this must be no less than the time from when the segment was first advertised to the time it takes to complete.
+
+For guidance on the Discontinuities, please see [Sequence Numbers](#sequence-numbers) and [Discontinuities](#ext-x-discontinuity).
+
+## General Client Responsibilities
+
+The client may opt into an LHLS stream. If so, the client must choose a prefetch Media Segment to play first from the Media Playlist when playback starts. The client must choose prefetch Media Segments for playback in the order in which they appear in the Playlist; however, the client may open connections to as many prefetch segments as desired. If data from a newer prefetch Media Segment is received before an older one, the client should not append this data to the SourceBuffer; doing so may stall playback. If the client opts out of LHLS, it must ignore all prefetch Media Segments, and any additional constraints outlined in this specification.
+
+The client may set a minimum amount of buffer to begin and maintain playback. The client should not impose a minimum buffered amount greater than one target duration; doing so may introduce undue latency.
```

Comment:
This feels like somewhere we might want to provide a little more guidance, smart defaults, or open the door to configure the target min buffer somehow. With ultra low latency there will be a fine balance between lower latency and more rebuffering, that will likely be audience dependent. A lot of players don't give you any buffer config options today.

Reply:
>clients should be able to build whatever experience best suits their usecase. 

Just to clarify this, they should build whatever *low-latency* usecase best suits them. The original intent behind this section was to ensure that they were operating like a low-latency player (and not just ignoring prefetch segments or having like 30s of buffer or whatever) but I don't think it came out right. I'll take another stab at this.

Reply:
+1 to target latency in the manifest.  There definitely needs to be a knob to turn for different use cases as they walk the line between latency and rebuffering

Reply:
I'd actually argue that the manifest should not dictate the playback latency. This may seem contrary to the discussion, but here are my reasons why:

1. Latency is a system feature, not dictated (or controlled) solely by the encoder. The appropriate latency should be a commercial target tempered by the characteristics of the client device, the last mile connectivity and the CDN. A mobile client on a 3G connection and an older mobile phone cannot sustain the same low latency as a fiber-connected SmartTV. They may be fed by different CDNs, yet all clients will receive the same playlist.
2. Architecturally the role of the playlist is to to describe the availability of media data. For service robustness and interoperability, this role should be cleanly encapsulated. It should not be overloaded to also carry application-level instructions such as which latency to play that content with, what screen size to show it on, whether to disallow skip during ads, which caption language to chose etc. Some clients should legitimately play the content 30s behind live, others 2s - it depends upon their use-case. All these service level decisions need to be carried out of band. When a client has to play media, it should be given a playlist (which describes the content availability) and some type of config/init file (which describes _how_ to play it). 

I think the manifest should add information to allow the latency to be estimated and I would support the spec saying that the package MUST add  #EXT-X-PROGRAM-DATE-TIME to the media playlists. What the various players to with this information is up to the service operator. 

Cheers
Will

Reply:
> I'd actually argue that the manifest should not dictate the playback latency. This may seem contrary to the discussion, but here are my reasons why:
> 
> 1. Latency is a system feature, not dictated (or controlled) solely by the encoder. The appropriate latency should be a commercial target tempered by the characteristics of the client device, the last mile connectivity and the CDN. A mobile client on a 3G connection and an older mobile phone cannot sustain the same low latency as a fiber-connected SmartTV. They may be fed by different CDNs, yet all clients will receive the same playlist.
I agree with you that the latency depends on a lot of factors. One of the key feature streaming service providers would want is that live playout to be synchronous across multiple clients in order to provide a common live experience for everybody, irrespective of their connection speed, device capabilites etc., It might be easier to solve this problem from the specification than everybody coming up with their own solutions.
Also a streaming service providers might want to have different latencies for different live events based on their use-case, contract obligations etc., Even in such cases it might be easier to control such things from encoder itself.
> 2. Architecturally the role of the playlist is to to describe the availability of media data. For service robustness and interoperability, this role should be cleanly encapsulated. It should not be overloaded to also carry application-level instructions such as which latency to play that content with, what screen size to show it on, whether to disallow skip during ads, which caption language to chose etc. Some clients should legitimately play the content 30s behind live, others 2s - it depends upon their use-case. All these service level decisions need to be carried out of band. When a client has to play media, it should be given a playlist (which describes the content availability) and some type of config/init file (which describes _how_ to play it).
The issue with an out-of-band config file for "how_to_play_it" is that everybody will come up with their own solutions which will lead to fragmentation within the industry. 
Even it is not mandatory/strict parameter, an optional parameter like "suggestedPresentationDelay" in DASH would be helpful for LHLS as well. 
> 
> I think the manifest should add information to allow the latency to be estimated and I would support the spec saying that the package MUST add #EXT-X-PROGRAM-DATE-TIME to the media playlists. What the various players to with this information is up to the service operator.
Yes, I agree. In any-case it is better if we make #EXT-X-PROGRAM-DATE-TIME mandatory for LHLS.
> 
> Cheers
> Will

Reply:
I agree with Will and his summary. The client will play the LHLS stream as best as it can given the current conditions and configuration. The problem with a target is that it's not actionable - if the client is behind because the network is slow, there's nothing it can do to get ahead. I was checking to see if DASH had something similar but couldn't find anything (but I didn't have the chance to look very hard). 

>But even then, manifest updates and network delays can cause us to stall if we try to play at the live edge. So we will actually delay where we will play at so we don't stall. The time we adjust by can be specified in the manifest using the MPD@suggestedPresentationDelay attribute. This specifies the delay to give the live stream to allow for smooth playback. If it isn't specified, we will give a reasonable default value.


`suggestedPresentationDelay` looks interesting (it seems to specify a _minimum_ latency), but I'm not sure how useful it is for LHLS. The encoder _must_ update the manifest with new segments on an interval equal to the average length of a segment.

>I think the manifest should add information to allow the latency to be estimated and I would support the spec saying that the package MUST add #EXT-X-PROGRAM-DATE-TIME to the media playlists.

I had #EXT-X-PROGRAM-DATE-TIME this in an original draft just for this purpose, but deleted it - i couldn't come up with an acceptable definition that all encoders could follow. Is the timestamp the time when the segment begins transcoding, has finished; or is it something like when the manifest was created in-memory (or between whatever stages in a transcoding pipeline). Maybe it doesn't matter too much. Input appreciated here, I'd like to add it back.

Reply:
There are indeed many potential reference points for where to hang the definition of #EXT-X-PROGRAM-DATE-TIME. MPEG has defined 6 if I remember correctly for the equivalent in DASH. There is no need for that complexity here. All you need is a COMMON reference point that all clients can access. They can then achieve synchronization (per @jkarthic-akamai comment above) by targeting a fixed delta from that point. A practical point would be the wallclock time at which that frame of media _entered_ the encoder. Assuming very small camera delay, the delta between that value and the wallclock time when the media frame is displayed by the client then represents the end-to-end latency. That's for lab confirmation. In the real production world, there is an unknown production delay upstream of the encoder (camera, OB truck, satellite contribution, broadcast profanity delay etc) so its very difficult for the end client to calculate the true e2e latency. Luckily, we don;t need to know the true e2e latency even for synch, we just need a consistent reference point between clients.

Reply:
> 
> I had #EXT-X-PROGRAM-DATE-TIME this in an original draft just for this purpose, but deleted it - i couldn't come up with an acceptable definition that all encoders could follow. Is the timestamp the time when the segment begins transcoding, has finished; or is it something like when the manifest was created in-memory (or between whatever stages in a transcoding pipeline). Maybe it doesn't matter too much. Input appreciated here, I'd like to add it back.

I agree that there is no need for us to define #EXT-X-PROGRAM-DATE-TIME strictly. We could just stick with the original definition from the official HLS spec. https://tools.ietf.org/html/draft-pantos-http-live-streaming-23#page-17 . We just suggest to make it as mandatory(MUST) parameter, so that all clients have a common reference point to sync.

Reply:
FWIW, we are relying on #EXT-X-PROGRAM-DATE-TIME to decide on what segment to start playback to reach our target latency. This approach has been successful for us.